### PR TITLE
Sk/pipeline ux ii

### PR DESCRIPTION
**Pull request by snopoke**
## Description
Some mostly small UI changes as well as some refactoring of the UI code.

Probably easiest reviewed by commit.

## User Impact
See demo.

### Demo
<div>
    <a href="https://www.loom.com/share/c8db54525071466799c1738ce61e24a7">
      <p>Changes to Pipelines Demo - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/c8db54525071466799c1738ce61e24a7">
      
    </a>
  </div>

### CODE REVIEW

**Comment by snopoke** [comment]
@coderabbit review

**Review by SmittieC — COMMENTED** [review]
There's just one comment about a zero division by coderabbit, otherwise looking good

File: assets/javascript/apps/pipeline/nodes/GetInputWidget.tsx [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+  MaxTokenLimitWidget,
+  SourceMaterialIdWidget,
+  TextWidget
+} from "../widgets";
+import React from "react";
+import {getCachedData} from "../utils";
+import {InputParam} from "../types/nodeInputTypes";
+import {NodeParams} from "../types/nodeParams";
+
+interface InputWidgetParams {
```
Comment (SmittieC) [inline assets/javascript/apps/pipeline/nodes/GetInputWidget.tsx:16]: Is there a particular reason you chose to use interface over a type for this?

Comment (SmittieC) [reply assets/javascript/apps/pipeline/nodes/GetInputWidget.tsx:16]: I did a bit of digging on when to use interface vs type aliases. [This article](https://www.totaltypescript.com/type-vs-interface-which-should-you-use) was insightful, but reaches a different conclusion to the typescript team ([typescript docs](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#differences-between-type-aliases-and-interfaces)). Interesting. Looks like we should be fine using either, but it's good to know the caveats mentioned in the first article

Comment (snopoke) [reply assets/javascript/apps/pipeline/nodes/GetInputWidget.tsx:16]: Thanks, that's useful. `type` definitely seems preferable in this instance.

**Comment by snopoke** [comment]
> There's just one comment about a zero division by coderabbit, otherwise looking good

I can't find this comment.

**Comment by snopoke** [comment]
> > There's just one comment about a zero division by coderabbit, otherwise looking good
> 
> I can't find this comment.

Resolved

File: assets/javascript/apps/pipeline/nodes/GetInputWidget.tsx (hunk 1 above)
Comment (snopoke) [reply assets/javascript/apps/pipeline/nodes/GetInputWidget.tsx]: 81c2b2ffa6c18648f49631d99d2166c1c3f8c864

**Comment by snopoke** [comment]
@coderabbit resolve

**Comment by snopoke** [comment]
@coderabbit pause

File: assets/javascript/apps/pipeline/nodes/GetOutputFactory.tsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    parseInt(
+      Array.isArray(params.num_outputs)
+        ? params.num_outputs.join("")
+        : params.num_outputs,
+    ) || 1;
+  const outputHandles = Array.from(
+    {length: numberOfOutputs},
+    (_, index) => {
+      const position = (index / (numberOfOutputs - 1)) * 100; // Distributes evenly between 0% to 100%
+      const handleAnnotation = handleText(`Output ${index + 1}`)
```
[the remark this answers is not in the document]
Comment (SmittieC) [reply assets/javascript/apps/pipeline/nodes/GetOutputFactory.tsx]: @snopoke here it is

**Review by SmittieC — APPROVED** [review]
Apologies, I didn't see you comment that 0 division is fixed :rocket:

**Pull request merged**